File: src/Ej3/Razas/Personaje.java
package Razas;

import java.util.HashMap;
import java.util.Map;

public class Personaje {
	protected String nombre, raza, clase;

	/* */
	private HashMap<String, Integer> primaryAttr;
	private HashMap<String, Integer> secondaryAttr;

	public Personaje() {
		primaryAttr = new HashMap<>();
		secondaryAttr = new HashMap<>();
	}

	public Personaje(String nom) {

		this.nombre = nom;
		this.raza = "";
		this.clase = "";
	}

	public HashMap<String, Integer> GetPrimary() {
		return primaryAttr;
	}

	public HashMap<String, Integer> GetSecondary() {
		return secondaryAttr;
	}

	public void SetPrimaryAttributes(HashMap<String, Integer> attr) {
		primaryAttr = attr;
	}

	public void SetSecondaryAttributes(HashMap<String, Integer> attr) {
		secondaryAttr = attr;
	}

	public void mostrarPersonaje() {
		System.out.println("\nPersonaje: " + nombre + ", " + this.clase + " " + this.raza + "\n\n"
				+ "-- Atributos primarios: \n"
				+ "----- Fuerza: " + this.primaryAttr.get("Fuerza") + "\n"
				+ "----- Destreza: " + this.primaryAttr.get("Destreza") + "\n"
				+ "----- Resistencia: " + this.primaryAttr.get("Resistencia") + "\n"
				+ "----- Inteligencia: " + this.primaryAttr.get("Inteligencia") + "\n"
				+ "----- Sabiduria: " + this.primaryAttr.get("Sabiduria") + "\n"
				+ "----- Carisma: " + this.primaryAttr.get("Carisma") + "\n"
				+ "----- Percepcion " + this.primaryAttr.get("Percepcion") + "\n\n"
				+ "-- Atributos secundarios: \n"
				+ "----- Vida: " + this.secondaryAttr.get("Vida") + "\n"
				+ "----- Estamina: " + this.secondaryAttr.get("Estamina") + "\n"
				+ "----- Mana: " + this.secondaryAttr.get("Mana") + "\n"
				+ "----- Persuasion: " + this.secondaryAttr.get("Persuasion") + "\n"
				+ "----- Agilidad: " + this.secondaryAttr.get("Agilidad") + "\n"
				+ "----- Intimidacion: " + this.secondaryAttr.get("Intimidacion") + "\n"
				+ "----- Critico: " + this.secondaryAttr.get("Critico") + "\n"
				+ "----- Punteria: " + this.secondaryAttr.get("Punteria") + "\n");
	}

	public void SetName(String nombre){
		this.nombre = nombre;
	}


	public void SetClass(String clase){
		this.clase = clase;
	}

	public void SetRace(String raza){
		this.raza = raza;
	}

}

File: Practica 3/ej4_creador_personajes/lib/Razas/Personaje.dart
import 'dart:collection';
import 'dart:io';

class Personaje {
  late String nombre, raza, clase;

  late HashMap<String, double> primaryAttr;
  late HashMap<String, double> secondaryAttr;

  Personaje(); //Dart no permite multiples constructores /overload

  //Estos constructores son como "de copia", y asi podemos tener varios constructores
  Personaje.hash(Personaje otro) {
    primaryAttr = HashMap();
    secondaryAttr = HashMap();
  }

  Personaje.nombre(Personaje otro, String nom) {
    nombre = nom;
    raza = "";
    clase = "";
  }

  HashMap<String, double> GetPrimary() {
    return primaryAttr;
  }

  HashMap<String, double> GetSecondary() {
    return secondaryAttr;
  }

  void SetPrimaryAttributes(HashMap<String, double> attr) {
    primaryAttr = attr;
  }

  void SetSecondaryAttributes(HashMap<String, double> attr) {
    secondaryAttr = attr;
  }

  void SetName(String nombre) {
    this.nombre = nombre;
  }

  void SetClass(String clase) {
    this.clase = clase;
  }

  void SetRace(String raza) {
    this.raza = raza;
  }

  void ExportState(String path) {
    // Abrir el archivo en modo escritura
    File file = File(path);
    IOSink sink = file.openWrite(mode: FileMode.write);

    // Vuelco las parejas key:value de atributos primarios
    primaryAttr.forEach((key, value) {
      sink.write('$key:$value\n');
    });

    // Vuelco las parejas key:value de atributos secundarios
    secondaryAttr.forEach((key, value) {
      sink.write('$key:$value\n');
    });

    // Cerrar el archivo
    sink.close();
  }
}
